events: Create `PossiblyRedacted` type for original state events

zecakeh opened the pull request:
Fix deserialization of redacted `prev_content`.

Fixes #1422.










<!-- Replace -->
----
Preview Removed
<!-- Replace -->

jplatte commented on crates/ruma-common/src/events/content.rs:
```diff
@@ -130,6 +130,9 @@ pub trait StateEventContent: EventContent<EventType = StateEventType> {
 pub trait OriginalStateEventContent: StateEventContent + RedactContent {
```
Now that we have them, should `RedactContent` members be "inlined" into the `Original` traits?

zecakeh replied:
We need `RedactContent` for `MessageLike` events, don't we?

jplatte replied:
Yes, we could add `type Redacted` and `fn redact` to both `OriginalStateEventContent` and `OriginalMessageLikeEventContent` and remove the trait, right? I don't think we have any code that is generic over `RedactContent`.

zecakeh replied:
Yes I think we could (although we need to create `OriginalMessageLikeEventContent).

Should I do this in this PR?

jplatte replied:
Doesn't have to be in this PR, no.

jplatte commented:
I would have expected most or all of the code here to be macro-generated, do you think that would work too?

zecakeh commented:
> I would have expected most or all of the code here to be macro-generated, do you think that would work too?

Yeah I could probably make more code generated by macro code by adding some attributes.

Do you also mean like building the custom `PossiblyRedacted` structs?

jplatte commented:
Would we need extra attributes? Couldn't we check whether there are any mandatory fields (not `Option` and no `serde(default)`) without `ruma_events(skip_redaction)`, and only then generate a `PossiblyRedacted` type with all those fields wrapped in `Option`? (and skip the whole thing for `custom_redacted` events)

zecakeh commented:
The thing I'm most afraid of are serde attributes. Even if right now I don't think we have any special case, someone using a `deserialize_with` for example would break if we change the type to an `Option`.

jplatte commented:
Right.. How about erroring out if there are any serde attributes on a field we would wrap in an `Option`?

zecakeh commented:
Sure, but then we probably need an attribute to deactivate the generation of the `PossiblyRedacted` type and do it manually.

jplatte commented:
Do you know we have any such types?

zecakeh commented:
I'm thinking [`m.policy.rule.*`](https://docs.ruma.io/ruma/events/policy/rule/index.html) types probably need a custom `PossiblyRedacted` implementation so they can be deserialized from an empty object (the current implementation in this PR is probably wrong, it looks like it will only deserialize from `null`, a missing `content` field in the event or an object with all the content fields). i.e. we need a `PossiblyRedacted` form of [`PolicyRuleEventContent`](https://docs.ruma.io/ruma/events/policy/rule/struct.PolicyRuleEventContent.html).

zecakeh committed:
events: Generate PossiblyRedacted type for original state events

Fix deserialization of redacted prev_content
Can be overriden with the `custom_possibly_redacted` attribute

zecakeh commented:
Now the code is generated by a macro when possible.

jplatte commented on crates/ruma-common/src/events/room/aliases.rs:
```diff
@@ -49,6 +49,11 @@ impl RedactContent for RoomAliasesEventContent {
+/// The possibly redacted form of [`RoomAliasesEventContent`].
+///
+/// This type is used when it's not obvious whether the content is redacted or not.
+pub type PossiblyRedactedRoomAliasesEventContent = RedactedRoomAliasesEventContent;
```
I think this this should be a separate type, in case extra fields that don't survive redaction are added to this event.

jplatte commented on crates/ruma-macros/src/events/event_content.rs:
```diff
@@ -452,6 +491,157 @@ fn generate_redacted_event_content<'a>(
+                let path: TypePath = parse_quote!{ Option<#old_type> };
+                attrs.push(parse_quote! { #[serde(skip_serializing_if = "Option::is_none")] });
+
+                Ok(Field { attrs, ty: path.into(), ..f.clone() })
```
```suggestion
                let ty = parse_quote!{ Option<#old_type> };
                attrs.push(parse_quote! { #[serde(skip_serializing_if = "Option::is_none")] });

                Ok(Field { attrs, ty, ..f.clone() })
```

jplatte commented on crates/ruma-macros/src/events/event_content.rs:
```diff
@@ -797,9 +990,22 @@ fn generate_static_event_content_impl(
+    // a generated possibly redacted struct also. If no `custom_redacted` or
+    // `custom_possibly_redacted` attrs are found the content needs a redacted
+    // struct generated.
+    !is_custom_redacted && !is_custom_possibly_redacted && event_kind == Some(EventKind::State)
```
```suggestion
    !is_custom_possibly_redacted && event_kind == Some(EventKind::State)
```

based on what I've seen above, everywhere we use `custom_redacted`, we could still actually use the generated `PossiblyRedacted` type, right?

zecakeh replied:
No, for `RoomMemberEventContent` we'd need a custom `PossiblyRedacted`. So in practice there's one case where it would work and one case where it wouldn't work.

jplatte replied:
Okay, let's keep the current implementation then I guess.

zecakeh replied:
You know what, I'm actually gonna change this. It's less annoying right now because I won't need to implement a custom struct and its traits. :stuck_out_tongue:

zecakeh committed:
Apply suggested change

Co-authored-by: Jonas Platte <jplatte@element.io>

zecakeh committed:
Generate PossiblyRedacted type for custom_redacted too

zecakeh committed:
Don't import TypePath
